feat: add C implementation for `stats/base/dists/gumbel/skewness`

ShabiShett07 opened the pull request:
Resolves #3613.

## Description

> What is the purpose of this pull request?

This pull request:

-   Solves the issue #3613 

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3613

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

ShabiShett07 committed:
feat: add /stats/base/dists/gumbel/skewness

ShabiShett07 committed:
fix: package.json

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

ShabiShett07 commented:
/stdlib lint-autofix

stdlib-bot committed:
fix: resolve lint errors

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/lib/native.js:
```diff
@@ -0,0 +1,62 @@
+
+/**
+* Returns the skewness for a Gumbel distribution with location `mu` and scale `beta`.
+*
+* @param {number} mu - location parameter
```
```suggestion
*
* @private
* @param {number} mu - location parameter
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/manifest.json:
```diff
@@ -0,0 +1,86 @@
+{
+    "options": {
```
Should have two-space indentation.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/src/main.c:
```diff
@@ -0,0 +1,67 @@
+/**
+* Returns the skewness for a Gumbel distribution with location `mu` and scale `beta`.
+*
+* @param {number} mu - location parameter
```
Should not be JSDoc but Doxygen; see other C implementations for how the comments should look like.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/src/main.c:
```diff
@@ -0,0 +1,67 @@
+* @param {PositiveNumber} beta - scale parameter
+* @returns {number} skewness
+*
+* @example
+* double y = stdlib_base_dists_gumbel_skewness( 0.0, 1.0 );
+* // returns ~1.14
+*
+* @example
+* double y = stdlib_base_dists_gumbel_skewness( 5.0, 2.0 );
+* // returns ~1.14
+*
+* @example
+* double y = stdlib_base_dists_gumbel_skewness( NaN, 1.0 );
+* // returns NaN
+*
+* @example
+* double y = stdlib_base_dists_gumbel_skewness( 0.0, NaN );
+* // returns NaN
+*
+* @example
+* double y = stdlib_base_dists_gumbel_skewness( 0.0, 0.0 );
+* // returns NaN
```
By convention, we only keep one example in the C comments:

```suggestion
* @example
* double y = stdlib_base_dists_gumbel_skewness( 0.0, 1.0 );
* // returns ~1.14
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/src/main.c:
```diff
@@ -0,0 +1,67 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+// MODULES //
+
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/special/sqrt.h"
+#include "stdlib/math/base/special/riemann_zeta.h"
+#include "stdlib/constants/float64/pi.h"
+
+
+// MAIN //
```
```suggestion
*/

#include "stdlib/math/base/assert/is_nan.h"
#include "stdlib/math/base/special/sqrt.h"
#include "stdlib/math/base/special/riemann_zeta.h"
#include "stdlib/constants/float64/pi.h"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/examples/c/example.c:
```diff
@@ -0,0 +1,40 @@
+	int i;
+
+	for ( i = 0; i < 25; i++ ) {
+		mu = random_uniform( 0.0, 10.0 ) - 5.0;
```
```suggestion
		mu = random_uniform( -5.0, 5.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,141 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		mu[ i ] = random_uniform( 0.0, 100.0 ) - 50.0;
+		beta[ i ] = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		mu[ i ] = random_uniform( -50.0, 50.0 );
		beta[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,71 @@
+	mu = new Float64Array( len );
+	beta = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		mu[ i ] = ( randu() * 100.0 ) - 50.0;
```
Would you be able to refactor this to draw directly from the desired distribution via `@stdlib/random/base/uniform`.

Planeshifter reviewed (changes requested):
Thanks @ShabiShett07!

Left a few comments; the `README.md` file will also have to be updated.

ShabiShett07 committed:
fix: update native.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
fix: removed multiple examples

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
fix: change example inputs

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
fix: change benchmark inputs

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
fix: main.c header files

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
fix: formayt errors

ShabiShett07 commented:
@Planeshifter, I have resolved all the conversations.

Planeshifter commented:
@ShabiShett07 Thank you! Only missing thing now is the C section for the `README.md` file.

ShabiShett07 committed:
chore: add c documentation

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/src/main.c:
```diff
@@ -0,0 +1,46 @@
+* // returns ~1.14
+*/
+double stdlib_base_dists_gumbel_skewness( const double mu, const double beta ) {
+	double SKEWNESS = 12.0 * stdlib_base_sqrt( 6.0 ) * stdlib_base_zeta( 3.0 ) / ( STDLIB_CONSTANT_FLOAT64_PI*STDLIB_CONSTANT_FLOAT64_PI*STDLIB_CONSTANT_FLOAT64_PI );
```
Noticing that this constant can be precomputed and moved to the outer scope as a `static const double`. This way, we can also avoid depending on the `sqrt` and `zeta` functions here. A similar change should be made to the JS implementation, if applicable.

ShabiShett07 replied:
Actually implementation has already been updated with such changes and I needed to make changes in c and I think it is ready

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/README.md:
```diff
@@ -133,6 +133,101 @@ for ( i = 0; i < 10; i++ ) {
+}
+
+int main( void ) {
+    double mu;
+    double beta;
+    double y;
+    int i;
```
```suggestion
    double beta;
    double mu;
    double y;
    int i;
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/examples/c/example.c:
```diff
@@ -0,0 +1,40 @@
+}
+
+int main( void ) {
+	double mu;
+	double beta;
+	double y;
+	int i;
```
```suggestion
	double beta;
	double mu;
	double y;
	int i;
```

Planeshifter reviewed (changes requested):
`README.md` looking good now, but just noticed we can precompute a constant.

ShabiShett07 committed:
chore: change variable format

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: change variable format

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: declare variable

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
@Planeshifter, Sorry there are some errors I will fix it in some minutes

ShabiShett07 committed:
chore: assigned constant value

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
@Planeshifter, ready for review

ShabiShett07 committed:
Merge branch 'develop' into feature/gumbel/skewness

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
fix: lint error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/src/main.c:
```diff
@@ -0,0 +1,49 @@
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/assert/is_nan.h"
+
+
+// VARIABLES //
+
+static const double SKEWNESS = 1.1395470994046488;
+
+
+// MAIN //
+
+/**
+* Returns the skewness for a Gumbel distribution with location `mu` and scale `beta`.
+*
+* @param mu     location parameter
+* @param beta   scale parameter
+* @returns      skewness
+*
+* @example
+* double y = stdlib_base_dists_gumbel_skewness( 0.0, 1.0 );
+* // returns ~1.14
+*/
```
```suggestion
#include "stdlib/math/base/assert/is_nan.h"

static const double SKEWNESS = 1.1395470994046488;

/**
* Returns the skewness for a Gumbel distribution with location `mu` and scale `beta`.
*
* @param mu     location parameter
* @param beta   scale parameter
* @return       skewness
*
* @example
* double y = stdlib_base_dists_gumbel_skewness( 0.0, 1.0 );
* // returns ~1.14
*/
```

Planeshifter committed:
style: remove headings and fix comment

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
chore: remove unused dependencies from manifest.json

Planeshifter reviewed (dismissed):
Thanks for so quickly addressing all comments! Let's merge this PR shortly.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/src/main.c:
```diff
@@ -0,0 +1,43 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/assert/is_nan.h"
```
```suggestion
#include "stdlib/stats/base/dists/gumbel/skewness.h"
#include "stdlib/math/base/assert/is_nan.h"
```

Planeshifter committed:
fix: add missing include

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/gumbel/skewness/src/main.c:
```diff
@@ -0,0 +1,43 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+#include "stdlib/stats/base/dists/gumbel/skewness.h"
+#include "stdlib/math/base/assert/is_nan.h"
```
```suggestion
*/

#include "stdlib/stats/base/dists/gumbel/skewness.h"
#include "stdlib/math/base/assert/is_nan.h"
```

Planeshifter committed:
style: add empty line

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
